implementation/jwt-auth/src/main/java/io/smallrye/jwt/auth/principal/PrincipalUtils.java

```diff
@@ -46,14 +46,23 @@ public static void setClaims(JwtClaims claimsSet, String token, JWTAuthContextIn
             String sub = findSubject(authContextInfo, claimsSet);
             claimsSet.setClaim(Claims.sub.name(), sub);
         }
-        Object groupsClaim = claimsSet.getClaimValue(Claims.groups.name());
-        if (groupsClaim == null || groupsClaim instanceof Map) {
-            List<String> groups = findGroups(authContextInfo, claimsSet);
-            claimsSet.setClaim(Claims.groups.name(), groups);
-        } else if (groupsClaim instanceof String) {
-            claimsSet.setClaim(Claims.groups.name(),
-                    splitStringClaimValue(groupsClaim.toString(), authContextInfo));
+
+        List<String> roles;
+        if (authContextInfo.getGroupsPath() == null) {
+            Object groupsClaim = claimsSet.getClaimValue(Claims.groups.name());
+            if (groupsClaim instanceof String) {
+                roles = splitStringClaimValue(groupsClaim.toString(), authContextInfo);
+            } else {
+                roles = List.class.cast(groupsClaim);
+            }
+        } else {
+            roles = findGroups(authContextInfo, claimsSet);
+        }
+
+        if (roles == null && authContextInfo.getDefaultGroupsClaim() != null) {
+            roles = Collections.singletonList(authContextInfo.getDefaultGroupsClaim());
         }
+        claimsSet.setClaim(Claims.groups.name(), roles);
```

Comment:
Please check if it is not null (the current code on the main branch does not check it but it should)

Reply:
This would then still return the `groups` claim although the custom path is set when the set custom path is empty.
This would mean this test case must be changed: https://github.com/smallrye/smallrye-jwt/pull/767/files#diff-30eaa73c4738eaad7b0a7e988717d1a2c0e5256141e3f00a3d7cc2cc8241db83R55
If that's what's desired, I'm okay with changing it...

I think it would be easier now that the tests are there to comment on those with the expected behavior.

Reply:
I assume you're talking about the `roles` or are you talking about the `claimsSet`?

Reply:
@oemel09 yes,`roles`

Reply:
Okay, I adapted the implementation